src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs

```diff
@@ -0,0 +1,473 @@
+        {
+            if (!this.isWritingJson)
+            {
+                this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+            }
+
+            this.textWriter?.Dispose();
+            this.Flush();
+
+            CheckIfSeparatorNeeded();
+            CheckIfManualValueAtArrayStart();
+        }
+
+        /// <summary>
+        /// Asynchronously starts a TextWriter value scope with the specified content type.
+        /// </summary>
+        /// <param name="contentType">The content type of the TextWriter value scope.</param>
+        /// <returns>A task representing the asynchronous operation. The task result is a TextWriter for writing JSON data.</returns>
+        public async Task<TextWriter> StartTextWriterValueScopeAsync(string contentType)
+        {
+            await this.FlushAsync().ConfigureAwait(false);
+
+            WriteItemWithSeparatorIfNeeded();
+
+            this.currentTextWriterContentType = contentType;
+
+            this.isWritingJson = string.IsNullOrEmpty(this.currentTextWriterContentType) || this.currentTextWriterContentType.StartsWith(MimeConstants.MimeApplicationJson, StringComparison.Ordinal);
+
+            if (!this.isWritingJson)
+            {
+                this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+            }
+
+            await this.FlushAsync().ConfigureAwait(false);
+
+            this.textWriter = new ODataUtf8JsonTextWriter(this);
+
+            return this.textWriter;
+        }
+
+        /// <summary>
+        /// Asynchronously ends a TextWriter value scope.
+        /// </summary>
+        /// <returns>A task representing the asynchronous operation.</returns>
+        public async Task EndTextWriterValueScopeAsync()
+        {
+            if (!this.isWritingJson)
+            {
+                this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+            }
+
+            if (this.textWriter != null)
+            {
+                await this.textWriter.DisposeAsync();
+            }
+
+            await this.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
+
+            CheckIfSeparatorNeeded();
+            CheckIfManualValueAtArrayStart();
+        }
+
+        /// <summary>
+        /// Represents a TextWriter implementation for writing UTF-8 JSON data to an ODataUtf8JsonWriter.
+        /// </summary>
+        internal sealed class ODataUtf8JsonTextWriter : TextWriter
+        {
+            private readonly ODataUtf8JsonWriter jsonWriter = null;
+            private char[] buffer;
+            private int numOfCharsNotWrittenFromPreviousChunk = 0;
+
+            public ODataUtf8JsonTextWriter(ODataUtf8JsonWriter jsonWriter)
+            {
+                this.jsonWriter = jsonWriter;
+            }
+
+            // Gets the character encoding in which the output is written.This is not supported by this text writer.
+            public override Encoding Encoding => throw new NotSupportedException();

```

Comment:
Why is returning `Encoding.UTF8;` not the right value to return given that this is a UTF8 writer?

Reply:
ODataUtf8JsonWriter uses the JavaScriptEncoder..